AIs won't bluff role that was challenged and will challenge role that was challenged successfully

einsteinsfool opened the pull request:
Another PR that passed all tests.
I will add more tests in the future after I write more PRs. For now AIs don't play as I'd expect them to so I prefer to first improve them.
I have over 10 more ideas for fixes/improvements.
And I finally know what's the difference between `state.playerIdx` and `state.state.playerIdx`. So it will be easier to change stuff now. :D

einsteinsfool commented:
As for adding the stringify, here's output from `game.js` without it:
```
players: 
 [ { name: 'knoob',
     cash: 2,
     influenceCount: 2,
     influence: [Object],
     isObserver: false,
     ai: false,
     isReady: true,
     connected: true },
```
And here is with:
```
"players": [
    {
        "name": "knoob",
        "cash": 2,
        "influenceCount": 2,
        "influence": [
            {
                "role": "assassin",
                "revealed": false
            },
            {
                "role": "duke",
                "revealed": false
            }
        ],
```

octachrome commented on ai-player.js:
```diff
@@ -249,6 +253,11 @@ function createAiPlayer(game, options) {
         }
 
+        // Challenge if somebody claimed this role and lost
+        if (calledBluffs[state.state.playerIdx] && calledBluffs[state.state.playerIdx][claimedRole]) {
```
I think this is not quite right: `state.state.playerIdx` is the player who took the initial action. If you are in ACTION_RESPONSE state, you need to check whether the *blocking player* (i.e., `state.state.target`) has been called on `claimedRole`.

octachrome commented on ai-player.js:
```diff
@@ -105,7 +105,7 @@ function createAiPlayer(game, options) {
                 playerIdx: state.state.target
             };
-        } else {
+        } else if (state.state.name != stateNames.REVEAL_INFLUENCE) {
```
I don't understand why this line needs to change - can you explain?

octachrome commented on ai-player.js:
```diff
@@ -512,12 +521,12 @@ function createAiPlayer(game, options) {
         // We are in reveal state.
```
I think maybe this whole function can be deleted - you are updating `calledBluffs` in `onHistoryEvent` instead, and that should be enough. Do you agree?

einsteinsfool commented:
Good remarks.
Added a comment why `lastRoleClaim` shouldn't be nulled on INFLUENCE_REVEAL state. It's because we still need this value to update `calledBluffs`.
You're right, I removed the unnecessary function since in `onHistoryEvent` we also update `calledBluffs` of AIs.
And finally, I think you meant:
> If you are in **BLOCK_RESPONSE** state, you need to check whether the blocking player
Fixed that too.

octachrome commented:
Good stuff!
